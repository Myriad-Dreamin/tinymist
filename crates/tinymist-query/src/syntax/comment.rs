use std::ops::Range;

use crate::prelude::*;
use crate::syntax::get_def_target;

fn extract_document_between(node: &LinkedNode, rng: Range<usize>) -> Option<String> {
    // collect all comments before the definition
    let mut comments = vec![];

    let mut newline_count = 0;
    let nodes = node.parent()?.children();
    for n in nodes {
        let offset = n.offset();
        if !rng.contains(&offset) {
            continue;
        }

        log::debug!("found comment for docs: {:?}: {:?}", n.kind(), n.text());

        match n.kind() {
            SyntaxKind::Hash => {
                newline_count = 0;
            }
            SyntaxKind::Space => {
                if n.text().contains('\n') {
                    newline_count += 1;
                }
                if newline_count > 1 {
                    comments.clear();
                }
            }
            SyntaxKind::Parbreak => {
                newline_count = 2;
                comments.clear();
            }
            SyntaxKind::LineComment => {
                newline_count = 0;
                // comments.push(n.text().strip_prefix("//")?.trim().to_owned());
                // strip all slash prefix
                let text = n.text().trim_start_matches('/');
                comments.push(text.trim().to_owned());
                continue;
            }
            SyntaxKind::BlockComment => {
                newline_count = 0;
                let text = n.text();
                let mut text = text.strip_prefix("/*")?.strip_suffix("*/")?.trim();
                // trip start star
                if text.starts_with('*') {
                    text = text.strip_prefix('*')?.trim();
                }
                comments.push(text.to_owned());
            }
            _ => {
                newline_count = 0;
            }
        }
    }

    if comments.is_empty() {
        return None;
    }

    Some(comments.join("\n"))
}

pub fn find_document_before(src: &Source, cursor: usize) -> Option<String> {
    log::debug!("finding docs at: {id:?}, {cursor}", id = src.id());

    let root = LinkedNode::new(src.root());
    let leaf = root.leaf_at(cursor)?;
    let def_target = get_def_target(leaf.clone())?;
    log::info!("found docs target: {:?}", def_target.node().kind());
    // todo: import
    let target = def_target.node().clone();
    let mut node = target.clone();
    while let Some(prev) = node.prev_sibling() {
        node = prev;
        if node.kind() == SyntaxKind::Hash {
            continue;
        }

        let start = node.range().end;
        let end = target.range().start;

        if end <= start {
            return None;
        }

        return extract_document_between(&node, start..end);
    }

    if node.parent()?.range() == root.range() && node.prev_sibling().is_none() {
        return extract_document_between(&node, root.offset()..node.range().start);
    }

    None
}
