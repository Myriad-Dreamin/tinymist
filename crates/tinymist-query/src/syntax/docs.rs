use std::{
    collections::BTreeMap,
    sync::{LazyLock, OnceLock},
};

use typst::foundations::{IntoValue, Module, Str, Type};

use crate::{
    adt::snapshot_map::SnapshotMap,
    analysis::SharedContext,
    docs::{
        convert_docs, identify_func_docs, identify_tidy_module_docs, identify_var_docs,
        DocStringKind, UntypedSymbolDocs, VarDocsT,
    },
    prelude::*,
    syntax::Decl,
    ty::{BuiltinTy, Interned, PackageId, SigTy, StrRef, Ty, TypeBounds, TypeVar, TypeVarBounds},
};

use super::DeclExpr;

/// The documentation string of an item
#[derive(Debug, Clone, Default)]
pub struct DocString {
    /// The documentation of the item
    pub docs: Option<EcoString>,
    /// The typing on definitions
    pub var_bounds: HashMap<DeclExpr, TypeVarBounds>,
    /// The variable doc associated with the item
    pub vars: BTreeMap<StrRef, VarDoc>,
    /// The type of the resultant type
    pub res_ty: Option<Ty>,
}

impl DocString {
    /// Get the documentation of a variable associated with the item
    pub fn get_var(&self, name: &StrRef) -> Option<&VarDoc> {
        self.vars.get(name)
    }

    /// Get the type of a variable associated with the item
    pub fn var_ty(&self, name: &StrRef) -> Option<&Ty> {
        self.get_var(name).and_then(|v| v.ty.as_ref())
    }
}

/// The documentation string of a variable associated with some item.
#[derive(Debug, Clone, Default)]
pub struct VarDoc {
    /// The documentation of the variable
    pub docs: EcoString,
    /// The type of the variable
    pub ty: Option<Ty>,
}

impl VarDoc {
    /// Convert the variable doc to an untyped version
    pub fn to_untyped(&self) -> Arc<UntypedSymbolDocs> {
        Arc::new(UntypedSymbolDocs::Variable(VarDocsT {
            docs: self.docs.clone(),
            return_ty: (),
            def_docs: OnceLock::new(),
        }))
    }
}

pub(crate) fn compute_docstring(
    ctx: &Arc<SharedContext>,
    fid: TypstFileId,
    docs: String,
    kind: DocStringKind,
) -> Option<DocString> {
    let checker = DocsChecker {
        fid,
        ctx,
        vars: HashMap::new(),
        globals: HashMap::default(),
        locals: SnapshotMap::default(),
        next_id: 0,
    };
    match kind {
        DocStringKind::Function => checker.check_func_docs(docs),
        DocStringKind::Variable => checker.check_var_docs(docs),
        DocStringKind::Module => checker.check_module_docs(docs),
        DocStringKind::Constant => None,
        DocStringKind::Struct => None,
        DocStringKind::Reference => None,
    }
}

struct DocsChecker<'a> {
    fid: TypstFileId,
    ctx: &'a Arc<SharedContext>,
    /// The typing on definitions
    vars: HashMap<DeclExpr, TypeVarBounds>,
    globals: HashMap<EcoString, Option<Ty>>,
    locals: SnapshotMap<EcoString, Ty>,
    next_id: u32,
}

impl<'a> DocsChecker<'a> {
    pub fn check_func_docs(mut self, docs: String) -> Option<DocString> {
        let converted = convert_docs(&self.ctx.world, &docs).ok()?;
        let converted = identify_func_docs(&converted).ok()?;
        let module = self.ctx.module_by_str(docs)?;

        let mut params = BTreeMap::new();
        for param in converted.params.into_iter() {
            params.insert(
                param.name.into(),
                VarDoc {
                    docs: param.docs,
                    ty: self.check_type_strings(&module, &param.types),
                },
            );
        }

        let res_ty = converted
            .return_ty
            .and_then(|ty| self.check_type_strings(&module, &ty));

        Some(DocString {
            docs: Some(converted.docs),
            var_bounds: self.vars,
            vars: params,
            res_ty,
        })
    }

    pub fn check_var_docs(mut self, docs: String) -> Option<DocString> {
        let converted = convert_docs(&self.ctx.world, &docs).ok()?;
        let converted = identify_var_docs(converted).ok()?;
        let module = self.ctx.module_by_str(docs)?;

        let res_ty = converted
            .return_ty
            .and_then(|ty| self.check_type_strings(&module, &ty.0));

        Some(DocString {
            docs: Some(converted.docs),
            var_bounds: self.vars,
            vars: BTreeMap::new(),
            res_ty,
        })
    }

    pub fn check_module_docs(self, docs: String) -> Option<DocString> {
        let converted = convert_docs(&self.ctx.world, &docs).ok()?;
        let converted = identify_tidy_module_docs(converted).ok()?;

        Some(DocString {
            docs: Some(converted.docs),
            var_bounds: self.vars,
            vars: BTreeMap::new(),
            res_ty: None,
        })
    }

    fn generate_var(&mut self, name: StrRef) -> Ty {
        self.next_id += 1;
        let encoded = Interned::new(Decl::Generated(DefId(self.next_id as u64)));
        log::debug!("generate var {name:?} {encoded:?}");
        let var = TypeVar {
            name,
            def: encoded.clone(),
        };
        let bounds = TypeVarBounds::new(var, TypeBounds::default());
        let var = bounds.as_type();
        self.vars.insert(encoded, bounds);
        var
    }

    fn check_type_strings(&mut self, m: &Module, strs: &str) -> Option<Ty> {
        let mut types = vec![];
        for name in strs.split(",").map(|e| e.trim()) {
            let Some(ty) = self.check_type_ident(m, name) else {
                continue;
            };
            types.push(ty);
        }

        Some(Ty::from_types(types.into_iter()))
    }

    fn check_type_ident(&mut self, m: &Module, name: &str) -> Option<Ty> {
        static TYPE_REPRS: LazyLock<HashMap<&'static str, Ty>> = LazyLock::new(|| {
            let values = Vec::from_iter(
                [
                    Value::None,
                    Value::Auto,
                    // Value::Bool(Default::default()),
                    Value::Int(Default::default()),
                    Value::Float(Default::default()),
                    Value::Length(Default::default()),
                    Value::Angle(Default::default()),
                    Value::Ratio(Default::default()),
                    Value::Relative(Default::default()),
                    Value::Fraction(Default::default()),
                    Value::Str(Default::default()),
                ]
                .map(|v| v.ty())
                .into_iter()
                .chain([
                    Type::of::<typst::visualize::Color>(),
                    Type::of::<typst::visualize::Gradient>(),
                    Type::of::<typst::visualize::Pattern>(),
                    Type::of::<typst::symbols::Symbol>(),
                    Type::of::<typst::foundations::Version>(),
                    Type::of::<typst::foundations::Bytes>(),
                    Type::of::<typst::foundations::Label>(),
                    Type::of::<typst::foundations::Datetime>(),
                    Type::of::<typst::foundations::Duration>(),
                    Type::of::<typst::foundations::Content>(),
                    Type::of::<typst::foundations::Styles>(),
                    Type::of::<typst::foundations::Array>(),
                    Type::of::<typst::foundations::Dict>(),
                    Type::of::<typst::foundations::Func>(),
                    Type::of::<typst::foundations::Args>(),
                    Type::of::<typst::foundations::Type>(),
                    Type::of::<typst::foundations::Module>(),
                ]),
            );

            let shorts = values
                .clone()
                .into_iter()
                .map(|ty| (ty.short_name(), Ty::Builtin(BuiltinTy::Type(ty))));
            let longs = values
                .into_iter()
                .map(|ty| (ty.long_name(), Ty::Builtin(BuiltinTy::Type(ty))));
            let builtins = [
                ("any", Ty::Any),
                ("bool", Ty::Boolean(None)),
                ("boolean", Ty::Boolean(None)),
                ("false", Ty::Boolean(Some(false))),
                ("true", Ty::Boolean(Some(true))),
            ];
            HashMap::from_iter(shorts.chain(longs).chain(builtins))
        });

        let builtin_ty = TYPE_REPRS.get(name).cloned();
        builtin_ty
            .or_else(|| self.locals.get(name).cloned())
            .or_else(|| self.check_type_annotation(m, name))
    }

    fn check_type_annotation(&mut self, m: &Module, name: &str) -> Option<Ty> {
        if let Some(v) = self.globals.get(name) {
            return v.clone();
        }

        let v = m.scope().get(name)?;
        log::debug!("check doc type annotation: {name:?}");
        if let Value::Content(c) = v {
            let annotated = c.clone().unpack::<typst::text::RawElem>().ok()?;
            let text = annotated.text().clone().into_value().cast::<Str>().ok()?;
            let code = typst::syntax::parse_code(&text.as_str().replace('\'', "Î¸"));
            let mut exprs = code.cast::<ast::Code>()?.exprs();
            let ret = self.check_type_expr(m, exprs.next()?);
            self.globals.insert(name.into(), ret.clone());
            ret
        } else {
            None
        }
    }

    fn check_type_expr(&mut self, m: &Module, s: ast::Expr) -> Option<Ty> {
        log::debug!("check doc type expr: {s:?}");
        match s {
            ast::Expr::Ident(i) => self.check_type_ident(m, i.get().as_str()),
            ast::Expr::FuncCall(c) => match c.callee() {
                ast::Expr::Ident(i) => {
                    let name = i.get().as_str();
                    match name {
                        "array" => Some({
                            let ast::Arg::Pos(pos) = c.args().items().next()? else {
                                return None;
                            };

                            Ty::Array(self.check_type_expr(m, pos)?.into())
                        }),
                        "tag" => Some({
                            let ast::Arg::Pos(ast::Expr::Str(s)) = c.args().items().next()? else {
                                return None;
                            };
                            let pkg_id = PackageId::try_from(self.fid).ok();
                            Ty::Builtin(BuiltinTy::Tag(Box::new((
                                s.get().into(),
                                pkg_id.map(From::from),
                            ))))
                        }),
                        _ => None,
                    }
                }
                _ => None,
            },
            ast::Expr::Closure(c) => {
                log::debug!("check doc closure annotation: {c:?}");
                let mut pos = vec![];
                let mut named = BTreeMap::new();
                let mut rest = None;
                let snap = self.locals.snapshot();

                let sig = None.or_else(|| {
                    for param in c.params().children() {
                        match param {
                            ast::Param::Pos(ast::Pattern::Normal(ast::Expr::Ident(i))) => {
                                let name = i.get().clone();
                                let base_ty = self.generate_var(name.as_str().into());
                                self.locals.insert(name, base_ty.clone());
                                pos.push(base_ty);
                            }
                            ast::Param::Pos(_) => {
                                pos.push(Ty::Any);
                            }
                            ast::Param::Named(e) => {
                                let exp = self.check_type_expr(m, e.expr()).unwrap_or(Ty::Any);
                                named.insert(e.name().into(), exp);
                            }
                            // todo: spread left/right
                            ast::Param::Spread(s) => {
                                let Some(i) = s.sink_ident() else {
                                    continue;
                                };
                                let name = i.get().clone();
                                let rest_ty = self.generate_var(name.as_str().into());
                                self.locals.insert(name, rest_ty.clone());
                                rest = Some(rest_ty);
                            }
                        }
                    }

                    let body = self.check_type_expr(m, c.body())?;
                    let sig = SigTy::new(pos.into_iter(), named, None, rest, Some(body)).into();

                    Some(Ty::Func(sig))
                });

                self.locals.rollback_to(snap);
                sig
            }
            ast::Expr::Dict(d) => {
                log::debug!("check doc dict annotation: {d:?}");
                None
            }
            _ => None,
        }
    }
}
