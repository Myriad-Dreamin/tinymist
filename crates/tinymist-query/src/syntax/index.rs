use std::str::FromStr;

use reflexo_typst::package::PackageSpec;
use rustc_hash::FxHashSet;

use crate::{adt::interner::Interned, prelude::*};

#[derive(Default)]
pub struct IndexInfo {
    pub(crate) paths: FxHashSet<Interned<str>>,
    pub(crate) packages: FxHashSet<PackageSpec>,
    pub(crate) identifiers: FxHashSet<Interned<str>>,
}

#[comemo::memoize]
pub fn get_index_info(src: &Source) -> Arc<IndexInfo> {
    let root = src.root();
    let mut worker = IndexWorker {
        info: IndexInfo::default(),
    };
    worker.visit(root);
    Arc::new(worker.info)
}

struct IndexWorker {
    info: IndexInfo,
}

impl IndexWorker {
    fn visit(&mut self, node: &SyntaxNode) {
        match node.cast::<ast::Expr>() {
            Some(ast::Expr::Str(path_str)) => {
                if path_str.to_untyped().text().len() > 65536 {
                    // skip long strings
                    return;
                }
                let path_str = path_str.get();

                if path_str.starts_with('@') {
                    let pkg_spec = PackageSpec::from_str(&path_str).ok();
                    if let Some(pkg_spec) = pkg_spec {
                        self.info.identifiers.insert(pkg_spec.name.clone().into());
                        self.info.packages.insert(pkg_spec);
                    }
                    return;
                }
                let path = Path::new(path_str.as_str());
                let name = path.file_name().unwrap_or_default().to_str();
                if let Some(name) = name {
                    self.info.paths.insert(name.into());
                }
            }
            Some(ast::Expr::MathIdent(ident)) => {
                self.info.identifiers.insert(ident.get().into());
            }
            Some(ast::Expr::Ident(ident)) => {
                self.info.identifiers.insert(ident.get().into());
            }
            _ => {}
        }

        for child in node.children() {
            self.visit(child);
        }
    }
}
