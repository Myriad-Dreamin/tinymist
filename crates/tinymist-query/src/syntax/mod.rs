//! Analyzing the syntax of a source file.
//!
//! This module must hide all **AST details** from the rest of the codebase.

#![allow(missing_docs)]

pub(crate) mod docs;
pub(crate) mod expr;
pub(crate) mod index;
pub(crate) mod lexical_hierarchy;
pub(crate) mod module;

pub(crate) use docs::*;
pub use expr::*;
pub use index::*;
pub use lexical_hierarchy::*;
pub use module::*;
pub use tinymist_analysis::syntax::*;
