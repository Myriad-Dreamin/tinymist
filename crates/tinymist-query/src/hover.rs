use core::fmt;

use typst_shim::syntax::LinkedNodeExt;

use crate::{
    analysis::{
        analyze_dyn_signature, find_definition, get_link_exprs_in, DefinitionLink, Signature,
    },
    jump_from_cursor,
    prelude::*,
    syntax::{find_docs_before, get_deref_target, LexicalKind, LexicalVarKind},
    ty::PathPreference,
    upstream::{expr_tooltip, plain_docs_sentence, route_of_value, truncated_repr, Tooltip},
    LspHoverContents, StatefulRequest,
};

/// The [`textDocument/hover`] request asks the server for hover information at
/// a given text document position.
///
/// [`textDocument/hover`]: https://microsoft.github.io/language-server-protocol/specification#textDocument_hover
///
/// Such hover information typically includes type signature information and
/// inline documentation for the symbol at the given text document position.
#[derive(Debug, Clone)]
pub struct HoverRequest {
    /// The path of the document to get hover information for.
    pub path: PathBuf,
    /// The position of the symbol to get hover information for.
    pub position: LspPosition,
}

impl StatefulRequest for HoverRequest {
    type Response = Hover;

    fn request(
        self,
        ctx: &mut AnalysisContext,
        doc: Option<VersionedDocument>,
    ) -> Option<Self::Response> {
        let doc_ref = doc.as_ref().map(|doc| doc.document.as_ref());

        let source = ctx.source_by_path(&self.path).ok()?;
        let offset = ctx.to_typst_pos(self.position, &source)?;
        // the typst's cursor is 1-based, so we need to add 1 to the offset
        let cursor = offset + 1;

        let contents = def_tooltip(ctx, &source, doc.as_ref(), cursor)
            .or_else(|| star_tooltip(ctx, &source, cursor))
            .or_else(|| link_tooltip(ctx, &source, cursor));

        let contents = contents.or_else(|| {
            Some(typst_to_lsp::tooltip(
                &ctx.tooltip(doc_ref, &source, cursor)?,
            ))
        })?;

        let ast_node = LinkedNode::new(source.root()).leaf_at_compat(cursor)?;
        let range = ctx.to_lsp_range(ast_node.range(), &source);

        // Neovim shows ugly hover if the hover content is in array, so we join them
        // manually with divider bars.
        let mut contents = match contents {
            LspHoverContents::Array(contents) => contents
                .into_iter()
                .map(|e| match e {
                    MarkedString::LanguageString(e) => {
                        format!("```{}\n{}\n```", e.language, e.value)
                    }
                    MarkedString::String(e) => e,
                })
                .join("\n\n---\n"),
            LspHoverContents::Scalar(MarkedString::String(contents)) => contents,
            LspHoverContents::Scalar(MarkedString::LanguageString(contents)) => {
                format!("```{}\n{}\n```", contents.language, contents.value)
            }
            lsp_types::HoverContents::Markup(e) => {
                match e.kind {
                    MarkupKind::Markdown => e.value,
                    // todo: escape
                    MarkupKind::PlainText => e.value,
                }
            }
        };

        if ctx.analysis.enable_periscope {
            if let Some(doc) = doc.clone() {
                let position = jump_from_cursor(&doc.document, &source, cursor);
                let position = position.or_else(|| {
                    for i in 1..100 {
                        let next_cursor = cursor + i;
                        if next_cursor < source.text().len() {
                            let position = jump_from_cursor(&doc.document, &source, next_cursor);
                            if position.is_some() {
                                return position;
                            }
                        }
                        let prev_cursor = cursor.checked_sub(i);
                        if let Some(prev_cursor) = prev_cursor {
                            let position = jump_from_cursor(&doc.document, &source, prev_cursor);
                            if position.is_some() {
                                return position;
                            }
                        }
                    }

                    None
                });

                log::info!("telescope position: {:?}", position);
                let content = position.and_then(|pos| ctx.resources.periscope_at(ctx, doc, pos));
                if let Some(preview_content) = content {
                    contents = format!("{preview_content}\n---\n{contents}");
                }
            }
        }

        Some(Hover {
            contents: LspHoverContents::Scalar(MarkedString::String(contents)),
            range: Some(range),
        })
    }
}

fn link_tooltip(
    ctx: &mut AnalysisContext<'_>,
    source: &Source,
    cursor: usize,
) -> Option<HoverContents> {
    let mut node = LinkedNode::new(source.root()).leaf_at_compat(cursor)?;
    while !matches!(node.kind(), SyntaxKind::FuncCall) {
        node = node.parent()?.clone();
    }

    let mut links = get_link_exprs_in(ctx, &node)?;
    links.retain(|link| link.0.contains(&cursor));
    if links.is_empty() {
        return None;
    }

    let mut results = vec![];
    let mut actions = vec![];
    for (_, target) in links {
        // open file in tab or system application
        actions.push(CommandLink {
            title: Some("Open in Tab".to_string()),
            command_or_links: vec![CommandOrLink::Command(Command {
                id: "tinymist.openInternal".to_string(),
                args: vec![JsonValue::String(target.to_string())],
            })],
        });
        actions.push(CommandLink {
            title: Some("Open Externally".to_string()),
            command_or_links: vec![CommandOrLink::Command(Command {
                id: "tinymist.openExternal".to_string(),
                args: vec![JsonValue::String(target.to_string())],
            })],
        });
        if let Some(kind) = PathPreference::from_ext(target.path()) {
            let preview = format!("A `{kind:?}` file.");
            results.push(MarkedString::String(preview));
        }
    }
    render_actions(&mut results, actions);
    if results.is_empty() {
        return None;
    }

    Some(LspHoverContents::Array(results))
}

fn star_tooltip(
    ctx: &mut AnalysisContext,
    source: &Source,
    cursor: usize,
) -> Option<HoverContents> {
    let leaf = LinkedNode::new(source.root()).leaf_at_compat(cursor)?;

    if !matches!(leaf.kind(), SyntaxKind::Star) {
        return None;
    }

    let mut leaf = &leaf;
    while !matches!(leaf.kind(), SyntaxKind::ModuleImport) {
        leaf = leaf.parent()?;
    }

    let mi: ast::ModuleImport = leaf.cast()?;
    let source = mi.source();
    let module = ctx.analyze_import(&leaf.find(source.span())?);
    log::debug!("star import: {source:?} => {:?}", module.is_some());

    let i = module?;
    let scope = i.scope()?;

    let mut results = vec![];

    let mut names = scope.iter().map(|(name, _, _)| name).collect::<Vec<_>>();
    names.sort();
    let mut items = String::new();
    items.push_str("let (");
    for name in &names {
        items.push_str(name);
        items.push_str(", ");
    }
    items.push_str(") = ..");

    results.push(MarkedString::LanguageString(LanguageString {
        language: "typc".to_owned(),
        value: items,
    }));

    Some(LspHoverContents::Array(results))
}

struct Command {
    id: String,
    args: Vec<JsonValue>,
}

enum CommandOrLink {
    Link(String),
    Command(Command),
}

struct CommandLink {
    title: Option<String>,
    command_or_links: Vec<CommandOrLink>,
}

fn def_tooltip(
    ctx: &mut AnalysisContext,
    source: &Source,
    document: Option<&VersionedDocument>,
    cursor: usize,
) -> Option<LspHoverContents> {
    let leaf = LinkedNode::new(source.root()).leaf_at_compat(cursor)?;

    let deref_target = get_deref_target(leaf.clone(), cursor)?;

    let lnk = find_definition(ctx, source.clone(), document, deref_target.clone())?;

    let mut results = vec![];
    let mut actions = vec![];

    match lnk.kind {
        LexicalKind::Mod(_)
        | LexicalKind::Var(LexicalVarKind::LabelRef)
        | LexicalKind::Var(LexicalVarKind::ValRef)
        | LexicalKind::Block
        | LexicalKind::Heading(..) => None,
        LexicalKind::Var(LexicalVarKind::Label) => {
            results.push(MarkedString::String(format!("Label: {}\n", lnk.name)));
            if let Some(c) = lnk.value.as_ref() {
                let c = truncated_repr(c);
                results.push(MarkedString::String(format!("{c}")));
            }
            Some(LspHoverContents::Array(results))
        }
        LexicalKind::Var(LexicalVarKind::BibKey) => {
            results.push(MarkedString::String(format!("Bibliography: @{}", lnk.name)));

            Some(LspHoverContents::Array(results))
        }
        LexicalKind::Var(LexicalVarKind::Function) => {
            let sig = if let Some(Value::Func(func)) = &lnk.value {
                Some(analyze_dyn_signature(ctx, func.clone()))
            } else {
                None
            };
            results.push(MarkedString::LanguageString(LanguageString {
                language: "typc".to_owned(),
                value: format!(
                    "let {name}({params});",
                    name = lnk.name,
                    params = ParamTooltip(sig)
                ),
            }));

            if let Some(doc) = DocTooltip::get(ctx, &lnk) {
                results.push(MarkedString::String(doc));
            }

            if let Some(link) = ExternalDocLink::get(ctx, &lnk) {
                actions.push(link);
            }

            render_actions(&mut results, actions);
            Some(LspHoverContents::Array(results))
        }
        LexicalKind::Var(LexicalVarKind::Variable) => {
            let deref_node = deref_target.node();
            // todo: check sensible length, value highlighting
            if let Some(values) = expr_tooltip(ctx.world(), deref_node) {
                match values {
                    Tooltip::Text(values) => {
                        results.push(MarkedString::String(values.into()));
                    }
                    Tooltip::Code(values) => {
                        results.push(MarkedString::LanguageString(LanguageString {
                            language: "typc".to_owned(),
                            value: values.into(),
                        }));
                    }
                }
            }

            results.push(MarkedString::LanguageString(LanguageString {
                language: "typc".to_owned(),
                value: format!("let {name};", name = lnk.name),
            }));

            if let Some(doc) = DocTooltip::get(ctx, &lnk) {
                results.push(MarkedString::String(doc));
            }

            if let Some(link) = ExternalDocLink::get(ctx, &lnk) {
                actions.push(link);
            }

            render_actions(&mut results, actions);
            Some(LspHoverContents::Array(results))
        }
    }
}

fn render_actions(results: &mut Vec<MarkedString>, actions: Vec<CommandLink>) {
    if actions.is_empty() {
        return;
    }

    let g = actions
        .into_iter()
        .map(|action| {
            // https://github.com/rust-lang/rust-analyzer/blob/1a5bb27c018c947dab01ab70ffe1d267b0481a17/editors/code/src/client.ts#L59
            let title = action.title.unwrap_or("".to_owned());
            let command_or_links = action
                .command_or_links
                .into_iter()
                .map(|col| match col {
                    CommandOrLink::Link(link) => link,
                    CommandOrLink::Command(command) => {
                        let id = command.id;
                        // <https://code.visualstudio.com/api/extension-guides/command#command-uris>
                        if command.args.is_empty() {
                            format!("command:{id}")
                        } else {
                            let args = serde_json::to_string(&command.args).unwrap();
                            let args = percent_encoding::utf8_percent_encode(
                                &args,
                                percent_encoding::NON_ALPHANUMERIC,
                            );
                            format!("command:{id}?{args}")
                        }
                    }
                })
                .collect::<Vec<_>>()
                .join(" ");
            format!("[{title}]({command_or_links})")
        })
        .collect::<Vec<_>>()
        .join(" | ");
    results.push(MarkedString::String(g));
}

// todo: hover with `with_stack`
struct ParamTooltip(Option<Signature>);

impl fmt::Display for ParamTooltip {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let Some(sig) = &self.0 else {
            return Ok(());
        };
        let mut is_first = true;
        let mut write_sep = |f: &mut fmt::Formatter<'_>| {
            if is_first {
                is_first = false;
                return Ok(());
            }
            f.write_str(", ")
        };

        let primary_sig = match sig {
            Signature::Primary(sig) => sig,
            Signature::Partial(sig) => {
                let _ = sig.with_stack;

                &sig.signature
            }
        };

        for p in &primary_sig.pos {
            write_sep(f)?;
            write!(f, "{}", p.name)?;
        }
        if let Some(rest) = &primary_sig.rest {
            write_sep(f)?;
            write!(f, "{}", rest.name)?;
        }

        if !primary_sig.named.is_empty() {
            let mut name_prints = vec![];
            for v in primary_sig.named.values() {
                name_prints.push((v.name.clone(), v.type_repr.clone()))
            }
            name_prints.sort();
            for (k, v) in name_prints {
                write_sep(f)?;
                let v = v.as_deref().unwrap_or("any");
                let mut v = v.trim();
                if v.starts_with('{') && v.ends_with('}') && v.len() > 30 {
                    v = "{ ... }"
                }
                if v.starts_with('`') && v.ends_with('`') && v.len() > 30 {
                    v = "raw"
                }
                if v.starts_with('[') && v.ends_with(']') && v.len() > 30 {
                    v = "content"
                }
                write!(f, "{k}: {v}")?;
            }
        }

        Ok(())
    }
}

struct ExternalDocLink;

impl ExternalDocLink {
    fn get(ctx: &mut AnalysisContext, lnk: &DefinitionLink) -> Option<CommandLink> {
        self::ExternalDocLink::get_inner(ctx, lnk)
    }

    fn get_inner(_ctx: &mut AnalysisContext, lnk: &DefinitionLink) -> Option<CommandLink> {
        if matches!(lnk.value, Some(Value::Func(..))) {
            if let Some(builtin) = Self::builtin_func_tooltip("https://typst.app/docs/", lnk) {
                return Some(builtin);
            }
        };

        lnk.value
            .as_ref()
            .and_then(|value| Self::builtin_value_tooltip("https://typst.app/docs/", value))
    }
}

impl ExternalDocLink {
    fn builtin_func_tooltip(base: &str, lnk: &DefinitionLink) -> Option<CommandLink> {
        let Some(Value::Func(func)) = &lnk.value else {
            return None;
        };

        use typst::foundations::func::Repr;
        let mut func = func;
        loop {
            match func.inner() {
                Repr::Element(..) | Repr::Native(..) => {
                    return Self::builtin_value_tooltip(base, &Value::Func(func.clone()));
                }
                Repr::With(w) => {
                    func = &w.0;
                }
                Repr::Closure(..) => {
                    return None;
                }
            }
        }
    }

    fn builtin_value_tooltip(base: &str, value: &Value) -> Option<CommandLink> {
        let base = base.trim_end_matches('/');
        let route = route_of_value(value)?;
        let link = format!("{base}/{route}");
        Some(CommandLink {
            title: Some("Open docs".to_owned()),
            command_or_links: vec![CommandOrLink::Link(link)],
        })
    }
}

pub(crate) struct DocTooltip;

impl DocTooltip {
    pub fn get(ctx: &mut AnalysisContext, lnk: &DefinitionLink) -> Option<String> {
        self::DocTooltip::get_inner(ctx, lnk).map(|s| "\n\n".to_owned() + &s)
    }

    fn get_inner(ctx: &mut AnalysisContext, lnk: &DefinitionLink) -> Option<String> {
        if matches!(lnk.value, Some(Value::Func(..))) {
            if let Some(builtin) = Self::builtin_func_tooltip(lnk) {
                return Some(plain_docs_sentence(builtin).into());
            }
        };

        let (fid, def_range) = lnk.def_at.clone()?;

        let src = ctx.source_by_id(fid).ok()?;
        find_docs_before(&src, def_range.start)
    }
}

impl DocTooltip {
    fn builtin_func_tooltip(lnk: &DefinitionLink) -> Option<&'_ str> {
        let Some(Value::Func(func)) = &lnk.value else {
            return None;
        };

        use typst::foundations::func::Repr;
        let mut func = func;
        let docs = 'search: loop {
            match func.inner() {
                Repr::Native(n) => break 'search n.docs,
                Repr::Element(e) => break 'search e.docs(),
                Repr::With(w) => {
                    func = &w.0;
                }
                Repr::Closure(..) => {
                    return None;
                }
            }
        };

        Some(docs)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::tests::*;

    #[test]
    fn test() {
        snapshot_testing("hover", &|world, path| {
            let source = world.source_by_path(&path).unwrap();

            let request = HoverRequest {
                path: path.clone(),
                position: find_test_position(&source),
            };

            let result = request.request(world, None);
            assert_snapshot!(JsonRepr::new_redacted(result, &REDACT_LOC));
        });
    }
}
