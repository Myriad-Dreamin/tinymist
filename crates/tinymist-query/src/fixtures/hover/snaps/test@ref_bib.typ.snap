---
source: crates/tinymist-query/src/hover.rs
expression: "JsonRepr::new_redacted(result, &REDACT_LOC)"
input_file: crates/tinymist-query/src/fixtures/hover/ref_bib.typ
---
{
 "contents": "Bibliography: `Russell:1908` [1]\n\n---\n\nB. Russell, Mathematical logic based on the theory of types, <span style=\"font-style: italic;\">American Journal of Mathematics</span>, 30, 222â€“262, 1908.",
 "range": "2:21:2:34"
}
